### DISCUSSION TITLE
https://joshuacook.netlify.app/posts/2024-07-27_python-data-analysis-org/

### DISCUSSION THREAD

**Discussion by giscus** [jhrcook/jhrcook-website#4] [Announcements]
# https://joshuacook.netlify.app/posts/2024-07-27_python-data-analysis-org/

This is a description of how I use Python in my data analyses, a system that has evolved over 7 years of a computational biology Ph.D. and working in industry.

https://joshuacook.netlify.app/posts/2024-07-27_python-data-analysis-org/

**Comment by 0ptimista** [comment]
Love this article!

Is there a template repo on github as reference?

**Reply by jhrcook** [reply]
Thank you for the kind words! It means a lot! 

I've been thinking of making a template for this system, but just haven't gotten around to it. Maybe I'll do it in the coming week – if I do, I'll post a reply here.

**Reply by 0ptimista** [reply]
Looking forward to it! 

Actually, I've been thinking this data analyse "system" for personal usage for a while. Your blog is a true sight for me! 

Really appericate!

**Comment by flyersworder** [comment]
Just wondering, do you also use Ruff to format notebooks? If not, do you have a way to do such a thing? I usually find that notebooks are quite messy and difficult to standardize. Thank you.

**Reply by jhrcook** [reply]
Thank you for pointing this out. Yes, I do use Ruff for formatting notebooks, too. I used to use black along with flake8, isort, and mypy though nbqa, so switching to ruff when it supported Jupyter notebooks natively became a no-brainer for me.

**Comment by a-alak** [comment]
Super interesting approach! 
Any repo in your GitHub that is implemented with this system, that we can check out for inspo?

**Reply by jhrcook** [reply]
Thank you! 
It's been on my to-do list for awhile now, but I haven't yet put together a repository. Since this post got so much traction and positive feedback, maybe I'll take care of it this weekend. I'll update this post if/when I do!

**Comment by amcdawes** [comment]
This came across my radar via the RealPython podcast, and what perfect timing. I have been inching toward a formal system like this and you've given me the inspiration to push it into reality. Great tips and +1 for a template (even if it's a totally silly analysis!)

**Reply by bbrewington** [reply]
Me too! Here’s the podcast link for anyone curious: https://realpython.com/podcasts/rpp/217/

Love the coverage of software engineering best practices for data people. We definitely need more of this (and software people have a lot to learn from data people 😁)

**Reply by jhrcook** [reply]
Thanks for the support and the heads up on the Podcast! Excited it's helping some people.

**Comment by bbrewington** [comment]
What did you mean by Docker being a walled garden? You might be referencing DockerHub, which is an online repository

If you provide a Dockerfile and optionally compose.yml, people can run that locally without having to mess w/ the hub (besides grabbing base images in FROM statements)

I get that all of that might be overkill for typical data analysis projects…was just curious

**Reply by jhrcook** [reply]
Thanks for pointing this out – it was a bit of an off-handed statement that I provided no support for 🙃. I meant more that it is more limiting for interactions with the system. For example, submitting jobs to a compute cluster or AWS, or accessing data on different volumes. Relatively simple tasks can become more onerous, requiring trial and error of tweaks to the Dockerfile. In general, I think it is a great tool, but perhaps more than is required for an analysis project. I think a virtual environment strikes a good balance of isolation (so I'll use 'uv' to make the virtual environment when I can, stepping up to 'conda'/'mamba' when I need to install more system-level packages). Docker feels like a bit too large of a hammer for most projects. If my thinking is wrong on this, I'm happy to be corrected!

(On our compute cluster, Docker access actually requires root access, meaning it would be possible to delete everything. This scared me enough to request removal from the access list haha.)

**Comment by moshejasper** [comment]
Just came across this. 

I have been increasingly drifting towards the same perspective as part of my bioinformatics PhD, and it is refreshing to see it so coherently named and articulated!
